lib/avro_turf/connection_wrapper_with_token_auth.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+class AvroTurf::ConnectionWrapperWithAuthToken < AvroTurf::ConnectionWrapper
+  REFRESH_TOKEN_TRIES = 4
+
+  def initialize(url,
+                 logger:,
+                 proxy: nil,
+                 user: nil,
+                 password: nil,
+                 ssl_ca_file: nil,
+                 client_cert: nil,
+                 client_key: nil,
+                 client_key_pass: nil,
+                 client_cert_data: nil,
+                 client_key_data: nil,
+                 oauth_url: nil,
+                 oauth_client_id: nil,
+                 oauth_client_secret: nil)
+    @oauth_url = oauth_url
+    @oauth_client_id = oauth_client_id
+    @oauth_client_secret = oauth_client_secret
+    @semaphore = Mutex.new
```

Comment:
I'm only halfway through reviewing this, so I need to see how this class is used, but this mutex is created per instance. Is this class used as a singleton? or are we trying to protect cross thread activity of the same instance? will come back to this.

Reply:
Yes, we maintain one instance of it when we use it, here: https://github.com/meetcleo/meetcleo/pull/23203/files#diff-46c130aaced44d9d3f3a294431826f310567402d64b4b388074ee7a0b9726828R20

Reply:
It's so we can take advantage of the caching, here: https://github.com/meetcleo/avro_turf/blob/master/lib/avro_turf/messaging.rb#L79

Reply:
perfect, I was actually just reviewing that now.